File: java/cours/src/session3/TD/MotIterable/Mot.java
package session3.TD.MotIterable;


interface Mot {
	default boolean estMotNonVide(){
		return false;
	}
	
	default boolean estConcat(){
		return false;
	}
	
	default char element(){
		throw new UnsupportedOperationException();
	}
	
	default Mot reste() {
		throw new UnsupportedOperationException();
	}
	
	default Mot gauche(){
		throw new UnsupportedOperationException();
	}
	
	default Mot droit(){
		throw new UnsupportedOperationException();
	}
	
	default Mot vide(){
		return MotVide.VIDE; 
	}
	
	default Mot mot(char n){
		return new MotNonVide(n, this);
	}
	
	default Mot concat(Mot ens){
		return new MotConcat(this, ens);
	}
	
	int taille();
	
	default boolean estVide(){
		return this.taille() == 0;
	}

}

File: typescript/session3/MotIterable/Mot.ts
import { Char } from "../Char";
import { MotNonVide } from './MotNonVide';
import { MotVide } from './MotVide';
import { MotConcat } from './MotConcat';

export abstract class Mot {
  estMotNonVide(): boolean {
    return false;
  }
  estConcat(): boolean {
    return false;
  }

  element(): Char {
    throw new Error();
  }
  reste(): Char {
    throw new Error();
  }

  gauche(): Mot {
    throw new Error();
  }

  droit(): Mot {
    throw new Error();
  }

  vide(): Mot {
    return MotVide.VIDE;
  }

  mot(n: Char): Mot {
    return new MotNonVide(n, this);
  }

  concat(ens: Mot): Mot {
    return new MotConcat(this, ens);
  }

  abstract taille(): number;

  estVide(): boolean {
    return this.taille() == 0;
  }
}